packages/services/src/setting/index.ts

```diff
@@ -42,11 +46,40 @@ export class SettingManager {
     const url = Private.url(base, id);
     const response = await makeRequest(url, {}, serverSettings);
 
+    if (!id) {
+      throw new Error('Plugin `id` parameter is required for settings fetch.');
+    }
+
     if (response.status !== 200) {
       throw new ResponseError(response);
     }
 
-    return response.json();
+    // Assert what type the server response is returning.
+    return response.json() as Promise<ISettingRegistry.IPlugin>;
+  }
+
+  /**
+   * Fetch the list of all plugin setting bundles.
+   *
+   * @returns A promise that resolves if successful.
+   */
+  async list(): Promise<{ ids: string[]; values: ISettingRegistry.IPlugin[] }> {
+    const { serverSettings } = this;
+    const { baseUrl, pageUrl } = serverSettings;
+    const { makeRequest, ResponseError } = ServerConnection;
+    const base = baseUrl + pageUrl;
+    const url = Private.url(base, '');
+    const response = await makeRequest(url, {}, serverSettings);
+
+    if (response.status !== 200) {
+      throw new ResponseError(response);
+    }
+
+    const json = await response.json();
+    const values: ISettingRegistry.IPlugin[] = (json || {})['settings'] || [];
```

Comment:
`ISettingRegistry.IPlugin` has a `data` attribute, but these `values` do not. The `data` attribute doesn't get set until https://github.com/jupyterlab/jupyterlab/pull/5470/files#diff-1926d857572cffe43160eec8a8e82828R189

I think we should either (a) not call these data values `ISettingRegistry.IPlugin` until that `.data` attribute is set, or (b) make the `.data` attribute optional.

Reply:
I think making the data optional would place too many burdens on consumers. So maybe we could do a variant of (a) and simply create a minimally empty `data` attribute right here when they are fetched.

Reply:
@afshin and I discussed this more at the dev meeting. I think we have a difference of opinion about how things should be structured, but neither of us feel super strongly one way or the other.

(a) One way of thinking about it is that an `IPlugin` has a populated `data` attribute with the user settings and the composed settings. Thus the things returned from the connector are *not* `IPlugin`s, but rather something more raw. The fetch transform takes in these raw values and transforms them into other raw things. I'm not sure when the compose data is calculated, but perhaps it is before the compose transform, and the compose transform takes and returns an `IPlugin`.

(b) Another way to think about it is that it is *all* dealing with an `IPlugin`, but an `IPlugin` may not have the `data` attribute populated until after the bootstrapping process is run. In that thinking, the fetch and compose transforms both operate on an `IPlugin`, but the `IPlugin` is in different stages of construction.

At this point, the easiest is to go with @afshin's new commit, which follows the (b) philosophy. I'm willing to try it out and see how it works in practice.